### PR TITLE
JAMES-3492

### PR DESCRIPTION
**Pull request by andrevka**
includes https://github.com/apache/james-project/pull/318
Use elasticsearch v7 modules everywhere

### CODE REVIEW

**Comment by mbaechler** [comment]
```
18:45:39,620 [INFO] Apache James :: Mailbox :: ElasticSearch ........... FAILURE [  8.768 s]
```
The module doesn't compile right now

File: pom.xml [hunk 1]
```diff
@@ -763,12 +763,23 @@
...
                 <artifactId>apache-james-mailbox-elasticsearch</artifactId>
                 <version>${project.version}</version>
                 <type>test-jar</type>
             </dependency>
+            <dependency>
+                <groupId>${james.groupId}</groupId>
+                <artifactId>apache-james-mailbox-elasticsearch-v7</artifactId>
+                <version>${project.version}</version>
+                <type>test-jar</type>
+            </dependency>
```
Comment (LanKhuat) [inline pom.xml:782]: Why do we still need the v6 version of some modules here?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 2]
```diff
@@ -108,15 +116,18 @@ default void getUsersQuotaShouldReturnAllUsersWhenNoParameters(WebAdminQuotaSear
...
-                .contentType(ContentType.JSON)
-                .body("username", containsInAnyOrder(
-                    BOB.asString(),
-                    JACK.asString(),
-                    GUY_WITH_STRANGE_DOMAIN.asString()));
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    given()
+                        .get("/quota/users")
+                        .then()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:123]: Indent

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 3]
```diff
@@ -127,16 +138,19 @@ default void getUsersQuotaShouldFilterOnDomain(WebAdminQuotaSearchTestSystem tes
...
-                .body("username", containsInAnyOrder(
-                    BOB.asString(),
-                    JACK.asString()));
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    given()
+                        .param("domain", LOST_LOCAL)
+                        .when()
+                        .get("/quota/users")
+                        .then()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:147]: Indent

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 4]
```diff
@@ -147,16 +161,19 @@ default void getUsersQuotaShouldLimitValues(WebAdminQuotaSearchTestSystem testSy
...
-                .body("username", containsInAnyOrder(
-                    BOB.asString(),
-                    GUY_WITH_STRANGE_DOMAIN.asString()));
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    given()
+                        .param("limit", 2)
+                        .when()
+                        .get("/quota/users")
+                        .then()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:170]: Indent

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 5]
```diff
@@ -167,16 +184,19 @@ default void getUsersQuotaShouldAcceptOffset(WebAdminQuotaSearchTestSystem testS
...
-                .body("username", containsInAnyOrder(
-                    JACK.asString(),
-                    GUY_WITH_STRANGE_DOMAIN.asString()));
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    given()
+                        .param("offset", 1)
+                        .when()
+                        .get("/quota/users")
+                        .then()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:193]: Indent

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 6]
```diff
@@ -190,16 +210,19 @@ default void getUsersQuotaShouldFilterOnMinOccupationRatio(WebAdminQuotaSearchTe
...
-                .body("username", containsInAnyOrder(
-                    JACK.asString(),
-                    GUY_WITH_STRANGE_DOMAIN.asString()));
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    given()
+                        .param("minOccupationRatio", 0.5)
+                        .when()
+                        .get("/quota/users")
+                        .then()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:219]: Indent

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 7]
```diff
@@ -213,16 +236,19 @@ default void getUsersQuotaShouldFilterOnMaxOccupationRatio(WebAdminQuotaSearchTe
...
-                .body("username", containsInAnyOrder(
-                    JACK.asString(),
-                    BOB.asString()));
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    given()
+                        .param("maxOccupationRatio", 0.5)
+                        .when()
+                        .get("/quota/users")
+                        .then()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:245]: Indent

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 8]
```diff
@@ -236,15 +262,18 @@ default void getUsersQuotaShouldOrderByUsername(WebAdminQuotaSearchTestSystem te
...
-                .contentType(ContentType.JSON)
-                .body("username", contains(
-                    BOB.asString(),
-                    GUY_WITH_STRANGE_DOMAIN.asString(),
-                    JACK.asString()));
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    given()
+                        .get("/quota/users")
+                        .then()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:269]: Indent

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java [hunk 9]
```diff
@@ -400,19 +429,21 @@ default void getUsersQuotaShouldReturnUserDetails(WebAdminQuotaSearchTestSystem
...
-                    .isEqualTo("[" +
+            CALMLY_AWAIT.atMost(Duration.ofSeconds(10))
+                .untilAsserted(() -> {
+                    String jsonAsString =
+                        with()
+                            .param("minOccupationRatio", 0.5)
+                            .get("/quota/users")
+                            .then()
+                            .statusCode(HttpStatus.OK_200)
+                            .extract()
```
Comment (LanKhuat) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:440]: Indent

**Review by LanKhuat — COMMENTED** [review]
Could you squash all the commits from #318 along with 0ee7303 commit?

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java [hunk 10]
```diff
@@ -47,6 +47,7 @@
 import org.elasticsearch.search.builder.SearchSourceBuilder;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Disabled;
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java:50]: Let's deprecate ES7 metric reporting and only rely on the prometheus endpoints, rather that some custom, unmaintained reporters?

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/JamesWithNonCompatibleElasticSearchServerTest.java [hunk 11]
```diff
@@ -47,10 +48,11 @@
 
     @AfterAll
     static void afterAll() {
-        dockerES2.stop();
+        dockerES6.stop();
     }
 
     @Test
+    @Disabled("JAMES-3492, fails because communication fails with lower than 7.x nodes, maybe should be removed ")
```
Comment (chibenwa) [inline server/container/guice/cassandra-guice/src/test/java/org/apache/james/JamesWithNonCompatibleElasticSearchServerTest.java:55]: Or we should adapt the test to show which failure is precisely happening?

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java (hunk 2 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java:123]: I will give a few more details of Lan comment that lacks precisions (habits)

When testing REST calls we align given/when/then/with at the same indent level to improve test readaiblity (like on the old diff).

You likely should preserve this indentation property, here and in other places.

**Comment by chibenwa** [comment]
I'm surprised JMAP integration tests did not break...

File: pom.xml (hunk 1 above)
Comment (andrevka) [reply pom.xml:782]: We do not. I will remove them

File: mailbox/elasticsearch/pom.xml [hunk 12]
```diff
@@ -34,11 +34,11 @@
     <dependencies>
         <dependency>
             <groupId>${james.groupId}</groupId>
-            <artifactId>apache-james-backends-es-v7</artifactId>
+            <artifactId>apache-james-backends-es</artifactId>
         </dependency>
         <dependency>
             <groupId>${james.groupId}</groupId>
-            <artifactId>apache-james-backends-es-v7</artifactId>
+            <artifactId>apache-james-backends-es</artifactId>
```
Comment (Arsnael) [inline mailbox/elasticsearch/pom.xml]: Why going back to ES6 in those bindings?

Comment (andrevka) [reply mailbox/elasticsearch/pom.xml]: it was just to get the project compiling. For now i just deleted those modules.

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java (hunk 10 above)
Comment (andrevka) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java:50]: It seems reasonable to me to rely solely on prometheus

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java (hunk 2 above)
Comment (andrevka) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserQuotaRoutesTest.java]: fixed

File: mailbox/elasticsearch/pom.xml (hunk 12 above)
Comment (Arsnael) [reply mailbox/elasticsearch/pom.xml]: Ok yes I realized I reviewed while you were pushing latest commits, sorry :)

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java (hunk 10 above)
Comment (andrevka) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java:50]: Shall i remove it from the project?

**Comment by andrevka** [comment]
Where should the elasticsearch update instructions be added?

**Comment by chibenwa** [comment]
In `/upgrade-instructions.md`

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java (hunk 10 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java:50]: Maybe we can raise the topic on server-dev mailing list, explain that it do not work with ES v6, that the underlying library is unmaintained (gasp) and that we have a pronmetheus alternative.

Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java:50]: I'll give my :+1: there.

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/JamesWithNonCompatibleElasticSearchServerTest.java (hunk 11 above)
Comment (andrevka) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/JamesWithNonCompatibleElasticSearchServerTest.java:55]: This stopped failing, probaly IDE was using wrong classes, will remove Disabled

File: server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java (hunk 10 above)
Comment (andrevka) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java:50]: I tried to add the question to the mailing list, but it did not work.

Comment (andrevka) [reply server/container/guice/cassandra-guice/src/test/java/org/apache/james/ESReporterTest.java:50]: https://www.mail-archive.com/server-dev@james.apache.org/msg69816.html

File: upgrade-instructions.md [hunk 13]
```diff
@@ -26,6 +26,34 @@ Change list:
...
 
+### Recommended upgrade to Elasticsearch 7.10.2
+
+Date: 11/03/2021
+
+JIRA: https://issues.apache.org/jira/browse/JAMES-2492
+
+Concerned products: Cassandra Guice server (with and without rabbitMQ)
+
+James is no longer tested against Elasticsearch 6.3.2 but instead against Elasticsearch 7.10.2 Users are recommended to upgrade to this
```
Comment (chibenwa) [inline upgrade-instructions.md:37]: `James is no longer tested against Elasticsearch 6.3.2`

Weknow it no longer is compatible, no?

**Comment by chibenwa** [comment]
Shall we rename packages and maven modules to remove the v7 suffix?

**Comment by chibenwa** [comment]
Nice job BTW!

**Comment by chibenwa** [comment]
We need a documentation update as well...

```
dockerfiles/run/docker-compose-branch-master.yml
dockerfiles/run/docker-compose.yml
README.adoc
docs/modules/servers/pages/distributed/run.adoc
docs/modules/servers/pages/distributed/operate/guide.adoc
docs/modules/servers/pages/distributed/operate/logging.adoc
docs/modules/servers/pages/distributed/run-docker.adoc
docs/modules/servers/pages/distributed/configure/elasticsearch.adoc
src/site/xdoc/server/quick-start-cassandra.xml
src/site/xdoc/server/config-elasticsearch.xml
src/site/markdown/server/install/guice-cassandra.md
src/site/markdown/server/install/guice-cassandra-rabbitmq-swift.md
```

File: upgrade-instructions.md (hunk 13 above)
Comment (andrevka) [reply upgrade-instructions.md:37]: > Shall we rename packages and maven modules to remove the v7 suffix?

I thought that maybe if in the future a new version change is made, then it is good to already have the v7 packages, so it is easier to create separate modules for example elasticsearch-v8 (then just need to rename the package not create them from null)

What do you think?

Comment (chibenwa) [reply upgrade-instructions.md:37]: I tend to agree with you, this sounds like a good explanation.

**Review by Arsnael — APPROVED** [review]
Nice work, huge thanks :)

**Review by chibenwa — APPROVED** [review]

**Review by LanKhuat — APPROVED** [review]

**Comment by chibenwa** [comment]
Hello!

Huge thanks for your work on this topic. I did just merge this PR.

I will then start working on the `3.6.0` release from tomorrow.

Cheers,

**Pull request closed without merging**